### PR TITLE
Support projections

### LINKED ISSUE #319: Adding PROJECTION to models
**Is your feature request related to a problem? Please describe.**
It would be awesome if we could add [projections](https://clickhouse.com/docs/en/sql-reference/statements/alter/projection) to models, this is very helpful when using a semantic layer on top of the data marts, to facilitate and speed up querying for certain filters.

**Describe the solution you'd like**
IMHO, the ideal solution would be to configure the projection as a `config` parameter (not sure though if dbt supports this):

```sql
-- models/marts/touchpoints;
{{
    config(
        materialized='materialized_view',
        engine='MergeTree()',
        order_by='workspace_id, toStartOfHour(updated_at), touchpoint_id',
        projections=[
            { "name": "some_projection", "select": ["touchpoint_id", "some_other_col"], "order_by": ["touchpoint_id"] }
        ]
    )
}}
select * from {{ source('other', 'touchpoints') }}
```

**Describe alternatives you've considered**
The alternative that we use currently is using a migration script to put the projection on various models, after they have been built, which is not really nice, since the code creating a projection and the model are separate from each other and we need to take care that the database migrations are always up to date with the dbt models.

### PR DESCRIPTION
**Pull request by BentsiLeviav**
## Summary
close #319 

This PR introduces support for projections in `dbt-clickhouse`. Projections are added to the `table` and `distributed_table` materializations as a model setting. For distributed tables, the projection is applied to the `_local` tables, not to the distributed proxy table.

## Tests
I added a test for each materialization. During test development, I encountered two challenges:

1. To verify that the projection was used by the query, we need to query `system.query_log`. However, this table takes time to create, which necessitates adding a sleep statement.
2. Currently, there's no straightforward way to pass a `query_id` parameter to ClickHouse, making it difficult to retrieve the exact query I want to check. As a workaround, I queried `system.query_log` using a `LIKE` statement.

### CODE REVIEW

File: dbt/include/clickhouse/macros/materializations/table.sql [hunk 1]
```diff
@@ -147,10 +147,28 @@
         {% call statement('create_table_empty') %}
             {{ create_table }}
         {% endcall %}
+         {% if config.get('projections') is not none%}
```
Comment (genzgd) [inline dbt/include/clickhouse/macros/materializations/table.sql:150]: In the previous file you don't have the "not none" piece.  I think it's better without, assuming the `if` evaluates as false for both `None` and empty arrays.

**Review by genzgd — APPROVED** [review]

File: tests/integration/adapter/projections/test_projections.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        result = project.run_sql(query, fetch="all")
+        assert len(result) == 3  # We expect 3 departments in the result
+        assert result == [('engineering', 43.666666666666664), ('malware', 40.0), ('sales', 25.0)]
+
+        # waiting for system.log table to be created
+        time.sleep(10)
+
+        # check that the latest query used the projection
+        result = project.run_sql(
+            f"SELECT query, projections FROM system.query_log WHERE query like '%{query}%' ORDER BY query_start_time DESC",
```
Comment (mshustov) [inline tests/integration/adapter/projections/test_projections.py:80]: clusterAllReplicas(default, 'system.query_log')

**Comment by BentsiLeviav** [comment]
Writing it here for community awareness - the feature was added both for `table` and `distributed_table` materialization, but it doesn't fully function with `distributed_table`. I marked the distributed test with `@pytest.mark.xfail`.

**Comment by arun11299** [comment]
@BentsiLeviav Would this be a problem when using Clickhouse Cloud ?

**Comment by BentsiLeviav** [comment]
@arun11299 it is supposed to work on ClickHouse Cloud with no problems.
For an example please refer to this changelog update
https://github.com/ClickHouse/dbt-clickhouse/blob/main/CHANGELOG.md#release-182-2024-08-22

**Comment by MarekVigas** [comment]
@BentsiLeviav I think it would be nice to at least mention it in README.md under [supported-features](https://github.com/ClickHouse/dbt-clickhouse?tab=readme-ov-file#supported-features) I really had to dig if projections are supported or not.

**Pull request merged**